### PR TITLE
Add convergence safeguards to threshold_li with oscillation detection and iteration limit

### PR DESCRIPTION
**Pull request by janniswi00**
## Description

This PR improves the robustness of threshold_li by adding a mechanism to detect and handle threshold oscillation during iteration. Specifically:

- A warning is now raised if the threshold value oscillates between already-seen values (e.g., two local optima).
- In the event of such oscillation, the iteration is explicitly terminated to avoid infinite loops.
- A maximum iteration limit (max_iter) was added to prevent infinite loops in edge cases.

Two tests were added:

- test_threshold_li_warns_on_oscillating_input to verify that oscillating inputs emit a UserWarning.
- test_threshold_li_breaks_at_max_iteration to check that the method stops after a given number of iterations and warns accordingly.

These additions help ensure graceful failure behavior in pathological cases and improve the method's robustness.

## Release note

```release-note
Improved `threshold_li` to detect oscillations and enforce a maximum number of iterations. Added user warnings in both cases.
```

### CODE REVIEW

**Comment by mkcor** [comment]
Closes #7790

**Review by jni — CHANGES_REQUESTED** [review]
Ach, I'm sorry that this slipped through the cracks @janniswi00! Thanks for this contribution!

I think we should get this in in some form, but for now there's a few issues:

1. I don't think we want to merge a 19MB test file into the repo, as every `git clone` forever will then fail. One idea is to add the image to our test data in gitlab.com/scikit-image/scikit-image, and then download it with pooch. See [Adding Data](https://scikit-image.org/docs/0.25.x/development/contribute.html#adding-data)
2. I think having a list of seen thresholds is too much work in each iteration. We only want to test the previous-to-last value, so we [should use a deque of maxlen 2](https://docs.python.org/3/library/collections.html#collections.deque).
3. And, since we know that t_curr and t_next are not close from the `while` condition, we can use `t_curr == (t_prev := seen[0])`
4. Note also: `round(t_curr, 10)` is an arbitrary level of precision, since folks could be looking at data in the range [10^-12, 10^-11]. I think that it's unlikely that the oscillation would continue unless the float values are *exactly* the same (but feel free to correct me on this), so I think `==` is enough. Alternatively, we should use the tolerance.

Let us know how this sounds! 🙏

**Comment by jni** [comment]
PS: As an alternative to adding the data file to our test data, I wonder whether it would be possible to construct a smaller pathological dataset from first principles, based on the update definition.

**Comment by jni** [comment]
> based on the update definition

here:

https://github.com/scikit-image/scikit-image/blob/0982d441d8ce4e873706309f5244f55e29a37dde/skimage/filters/thresholding.py#L771

File: skimage/filters/thresholding.py [hunk 1]
```diff
@@ -772,6 +784,17 @@ def threshold_li(image, *, tolerance=None, initial_guess=None, iter_callback=Non
 
             if iter_callback is not None:
                 iter_callback(t_next + image_min)
+            
+
+            if round(float(t_curr), 10) in seen_thresholds:
+                warnings.warn("threshold_li oscillates between two threshold values", UserWarning)
```
Comment (lagru) [inline skimage/filters/thresholding.py:790]: A [`RuntimeWarning`](https://docs.python.org/3/library/exceptions.html#RuntimeWarning) might fit here too.

cc @stefanv

**Comment by jni** [comment]
> wonder whether it would be possible to construct a smaller pathological dataset from first principles

fwiw, I have tried and failed at this. Here was my approach:

- Take the pathological image from the example, and the two thresholds that Li oscillates between, t0 and t1.

- Imagine the image with values sorted by whether they are below t0, between t0 and t1, and above t1. Additionally, at the very beginning of the array, you have the minimum value, z:

```
z, a0, a1, a2, ..., an, b0, b1, b2, ... bm, c0, c1, c2, ..., cm
                       ^                   ^
                       |                   |
                       t0                  t1
```

(It turns out that there is exactly one `b` value between t0 and t1, but that's not important for our purposes.)

So, when the current threshold is t0, the background and foreground means are `µ00 = mean(z, a0, ..., an)` and `µ01 = mean(b0, ..., cm)`, and when the threshold is t1, the means are `µ10 = mean(z, a0, ..., bm)` and `µ11 = mean(c0, ..., cm)`. 

So, how do we construct a simple image for which the iteration values are the same? Well, we can replace each segment with its mean, repeated the same number of times as the original values. If we let `µb = mean(b0, ..., bm)`, we can construct this image:

```
µ00, ... (n+1 times), µb, ... (m times), µ11, ... (p times)
```

then means used in the iteration will be unchanged.

- ... except, because of numerical precision issues, that's not the case. So actually in this case, (as in the float64 case with the exact same data) just terminates innocently.

- I also noted that the [minimum is subtracted](https://github.com/scikit-image/scikit-image/blob/0982d441d8ce4e873706309f5244f55e29a37dde/skimage/filters/thresholding.py#L713-L714) from the data before doing the iteration (to ensure that the mean values strictly greater than 0), so I tried to instead construct:

```
z, µa (n times), µb, ... (m times), µ11, ... (p times)
```

But that still terminated innocently! Ultimately, looking at the behaviour in detail, it relies on some very specific numerical computation errors (which is why the same image with float64 doesn't cause an infinite loop).

In the end, I'm inclined to test for 2-element loops (rather than checking any previously seen value), have a maximum iteration limit, *and leave the checks untested*, because it's such a weird edge case. You can test the iteration limit with a normal image and a very low limit. (ie, I would just drop the test data and address points 2-4 in [my earlier post](https://github.com/scikit-image/scikit-image/pull/7798#pullrequestreview-2909379164).

**Comment by janniswi00** [comment]
@stefanv yes I'll take it on. I'll implement the changes @jni proposed.

**Comment by matthew-brett** [comment]
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
